go-controller/pkg/ovn/ovn.go

```diff
@@ -59,23 +61,63 @@ type ACLLoggingLevels struct {
 	Deny  string `json:"deny,omitempty"`
 }
 
-// Controller structure is the object which holds the controls for starting
-// and reacting upon the watched resources (e.g. pods, endpoints)
-type Controller struct {
+// GenericController structure is place holder for all fields shared among controllers.
+type GenericController struct {
 	client       clientset.Interface
 	kube         kube.Interface
 	watchFactory *factory.WatchFactory
-	stopChan     <-chan struct{}
+	podRecorder  metrics.PodRecorder
 
-	// FIXME DUAL-STACK -  Make IP Allocators more dual-stack friendly
-	masterSubnetAllocator        *subnetallocator.SubnetAllocator
-	hybridOverlaySubnetAllocator *subnetallocator.SubnetAllocator
+	// event recorder used to post events to k8s
+	recorder record.EventRecorder
+
+	// libovsdb northbound client interface
+	nbClient libovsdbclient.Client
+
+	// libovsdb southbound client interface
+	sbClient libovsdbclient.Client
 
+	// has SCTP support
 	SCTPSupport bool
+}
+
+// ControllerManager structure is the object manages all controllers for all networks
+type ControllerManager struct {
```

Comment:
Unless I am mistaken, this ControllerManager is the thing that's is going to monitor NADs and spawn other OvnController(s), right?. I don't think ControllerManager needs to be in this `ovn.go` file at all, as it has a very clear and independent mission different than OvnController's one; and this file is already big enough and keeps getting things thrown an it. In the same vein I don't think we need both ControllerManager and OvnController embed a common type GenericController just because they share some attributes. They are conceptually different things in charge of different tasks.

In my opinion, we need to have this ControllerManager somewhere else, possibly even a different package. The independence of ControllerManager with OvnController needs to be laid out with stronger emphasis, it is just confusing to have stuff that is doing different things so intertwined. But this might be against what @trozet suggested, so lets see what he thinks.

I would also like to see a bit more on the interface of how ControllerManager is going to manage other OvnController(s), not only the default. I see more specific things for the default OvnController in ControllerManager that I would expect, and fewer things than  what I would expect to manage other seconday OvnController(s). Otherwise I can't really say much because I don' know what  to expect.

@maiqued @trozet WDYT?

Reply:
> Unless I am mistaken, this ControllerManager is the thing that's is going to monitor NADs and spawn other OvnController(s), right?. I don't think ControllerManager needs to be in this `ovn.go` file at all, as it has a very clear and independent mission different than OvnController's one; and this file is already big enough and keeps getting things thrown an it. In the same vein I don't think we need both ControllerManager and OvnController embed a common type GenericController just because they share some attributes. They are conceptually different things in charge of different tasks.

I completely agree with the first point in this paragraph - I think the `ControllerManager` should go to a different - multi-homing related ?? - package. 

I don't agree as much with the second - but I do think the name of the struct doesn't reflect well what that struct does.

AFAIU, the `GenericController` does nothing - it is simply bundling together a bunch of clients to k8s / OVN / etc. 

I think renaming it to something more accurate would be good or even splitting it by responsibilities (K8s API / OVN API / ...) would make it easier to follow. 

> 
> In my opinion, we need to have this ControllerManager somewhere else, possibly even a different package. The independence of ControllerManager with OvnController needs to be laid out with stronger emphasis, it is just confusing to have stuff that is doing different things so intertwined. But this might be against what @trozet suggested, so lets see what he thinks.
> 
> I would also like to see a bit more on the interface of how ControllerManager is going to manage other OvnController(s), not only the default. I see more specific things for the default OvnController in ControllerManager that I would expect, and fewer things than what I would expect to manage other seconday OvnController(s).

I also agree with this last sentence.

> 
> @maiqued @trozet WDYT?

Reply:
yeah I agree with the sentiments here. I wasn't sure how much separation there is between the cluster manager and an ovn controller when first reviewing, but I think having the 2 things separated with an interface defining what the cluster manager does and removing things it doesn't need is beneficial.